hack/tools/prowjob-gen/generator.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+)
+
+const generatedFileHeader = "# Code generated by cluster-api's prowjob-gen. DO NOT EDIT.\n"
+
+// newGenerator initializes a generator which includes parsing the configured templates.
+func newGenerator(config Config, templatesDir, outputDir string) (*generator, error) {
+	g := &generator{
+		config:       config,
+		outputDir:    outputDir,
+		createdFiles: map[string]bool{},
+	}
+
+	var err error
+	g.templates, err = template.New("").
+		Funcs(g.templateFunctions()).
+		ParseGlob(templatesDir + "/*.yaml.tpl")
+	if err != nil {
+		return nil, err
+	}
+
+	return g, err
+}
+
+type generator struct {
+	templates    *template.Template
+	config       Config
+	outputDir    string
+	createdFiles map[string]bool
+}
+
+// generate executes every template for every branch and writes the result to a
+// file in outputDir.
+func (g *generator) generate() error {
+	for _, tpl := range g.config.Templates {
+		for branch := range g.config.Branches {
+			out, err := g.executeTemplate(branch, tpl.Name)
+			if err != nil && !errors.Is(err, os.ErrNotExist) {
+				return errors.Wrapf(err, "Generating prowjobs for template %s", tpl.Name)
+			}
+
+			fileName := fmt.Sprintf(tpl.Format, strings.ReplaceAll(branch, ".", "-"))
+			filePath := filepath.Clean(path.Join(g.outputDir, fileName))
+			if err := os.WriteFile(filePath, out.Bytes(), 0644); err != nil { //nolint:gosec
+				return errors.Wrapf(err, "Writing prowjob to %q", filePath)
+			}
+
+			g.createdFiles[fileName] = true
+		}
+	}
+	return nil
+}
+
+// cleanup deletes files which have the generatedFileHeader and had not been updated
```

Comment:
Wondering if we should just delete all files with the header and simply do it before we write the new files